### ISSUE TITLE
Vector aggregator memory leaks

### ISSUE THREAD

**Issue by namm2** [vectordotdev/vector#9999]
### Vector Version

```
v0.16.1, then v0.17.3
```

### Vector Configuration File

```yaml
sinks:
 elk:
    doc_type: doc
    encoding:
      timestamp_format: rfc3339
    endpoint: http://elasticsearch-client.namespace.svc:9200
    healthcheck: ${ELK_SINK_HEALTH_CHECK:-true}
    index: logstash-%Y.%m.%d
    inputs:
    - transformer
    type: elasticsearch
sources:
  internal_logs:
    type: internal_logs
  fluent_input:
    address: 0.0.0.0:9090
    type: fluent
```

### Additional Context

Hi, we're running Vector-aggregator stateful set in GKE to process GCP load balancer logs, the topology looks like:

```[ Cloud Logging ] ---> [ Pub/Sub ]  ---> [ Fluentd STS] ---> [ Vector Aggregator ]```

Vector's configs are close to default settings as posted above.

Occasionally, one vector-aggregator pod consumes a lot of memory (and CPU is also running flat ~ 4 vCPUs) until it reached node's allocatable then got evicted. Afterwards, it's running fine for awhile for unknown duration and the issue happens again.

<img width="1292" alt="Screen Shot 2021-11-11 at 08 41 28" src="https://user-images.githubusercontent.com/46129063/141259605-1d6cbcd4-b347-4ba2-af73-ad5cfe4a0bb7.png">

**Comment by mrnonz** [comment]
How about this bug? Did you still face on v0.23.3?

**Comment by namm2** [comment]
@mrnonz I don't see this bug anymore on v0.23

**Comment by jszwedko** [comment]
Closing since this seems to be resolved in v0.23, but please feel free to leave a comment or open a new issue if you still see this behavior.
